transaction/std/src/transaction_builder.rs

```diff
@@ -2595,4 +2779,333 @@ pub mod transaction_builder_tests {
+                .prefix
+                .outputs
+                .iter()
+                .position(|tx_out| tx_out.public_key == tx_out1.public_key)
+                .unwrap();
+            let idx2 = tx
+                .prefix
+                .outputs
+                .iter()
+                .position(|tx_out| tx_out.public_key == tx_out2.public_key)
+                .unwrap();
+            let change_idx = (0..3).find(|x| *x != idx1 && *x != idx2).unwrap();
+
+            let change_tx_out = &tx.prefix.outputs[change_idx];
+
+            // Test that sender's change subaddress owns the change, and not the other tx
+            // outs
+            assert!(
+                !subaddress_matches_tx_out(&sender, CHANGE_SUBADDRESS_INDEX, &tx_out1).unwrap()
+            );
+            assert!(
+                !subaddress_matches_tx_out(&sender, CHANGE_SUBADDRESS_INDEX, &tx_out2).unwrap()
+            );
+            assert!(
+                subaddress_matches_tx_out(&sender, CHANGE_SUBADDRESS_INDEX, &change_tx_out)
+                    .unwrap()
+            );
+
+            // Test that recipients's default subaddress owns the change, and not the other
+            // tx outs
+            assert!(
+                subaddress_matches_tx_out(&recipient, DEFAULT_SUBADDRESS_INDEX, &tx_out1).unwrap()
+            );
+            assert!(
+                subaddress_matches_tx_out(&recipient, DEFAULT_SUBADDRESS_INDEX, &tx_out2).unwrap()
+            );
+            assert!(!subaddress_matches_tx_out(
+                &recipient,
+                DEFAULT_SUBADDRESS_INDEX,
+                &change_tx_out
+            )
+            .unwrap());
+
+            // Test that view key matching works with the two tx outs
+            let (amount, _) = tx_out1
+                .view_key_match(&recipient.view_private_key())
+                .unwrap();
+            assert_eq!(
+                amount.value,
+                amount1.value - change_amount.value - Mob::MINIMUM_FEE
+            );
+            assert_eq!(amount.token_id, Mob::ID);
+
+            let (amount, _) = tx_out2
+                .view_key_match(&recipient.view_private_key())
+                .unwrap();
+            assert_eq!(amount, amount2);
+
+            assert!(change_tx_out
+                .view_key_match(&recipient.view_private_key())
+                .is_err());
+
+            // Test that view key matching works with the change tx out with sender's view
+            // key
+            let (amount, _) = change_tx_out
+                .view_key_match(&sender.view_private_key())
+                .unwrap();
+            assert_eq!(amount.value, change_amount.value);
+
+            assert!(tx_out1.view_key_match(&sender.view_private_key()).is_err());
+
+            assert!(tx_out2.view_key_match(&sender.view_private_key()).is_err());
+        }
+    }
+
+    #[test]
+    // Test mixed transactions expected failure case 1
+    fn test_mixed_transactions_expected_failure1() {
```

Comment:
What is the expected failure?

Reply:
i subtracted 1 from one of the outputs, so that it isn't balanced

Reply:
will document this a little better